codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/protocols/serialize/AwsQuerySerializerGenerator.kt

```diff
@@ -1,56 +1,304 @@
         "Error" to serializerError,
         "SdkBody" to RuntimeType.sdkBody(runtimeConfig),
+        "QueryWriter" to smithyQuery.member("QueryWriter"),
+        "QueryValueWriter" to smithyQuery.member("QueryValueWriter"),
     )
 
+    override fun documentSerializer(): RuntimeType {
+        TODO("AwsQuery doesn't support document types")
+    }
+
     override fun payloadSerializer(member: MemberShape): RuntimeType {
-        val fnName = symbolProvider.serializeFunctionName(member)
-        val target = model.expectShape(member.target, StructureShape::class.java)
+        TODO("The Aws Query protocol doesn't support http payload traits")
+    }
+
+    override fun operationSerializer(operationShape: OperationShape): RuntimeType? {
+        val fnName = symbolProvider.serializeFunctionName(operationShape)
+        val inputShape = operationShape.inputShape(model)
         return RuntimeType.forInlineFun(fnName, "operation_ser") { writer ->
             writer.rustBlockTemplate(
                 "pub fn $fnName(input: &#{target}) -> Result<#{SdkBody}, #{Error}>",
-                *codegenScope,
-                "target" to symbolProvider.toSymbol(target)
+                *codegenScope, "target" to symbolProvider.toSymbol(inputShape)
             ) {
-                // TODO: Implement query payload serializer
-                writer.rust("unimplemented!()")
+                val action = operationShape.id.name
+                val version = serviceShape.version
+                rust("let mut out = String::new();")
+                rustTemplate(
+                    "let mut writer = #{QueryWriter}::new(&mut out, ${action.dq()}, ${version.dq()});",
+                    *codegenScope
+                )
+                serializeStructure(Context("writer", ValueExpression.Reference("input"), inputShape), true)
+                rust("writer.finish();")
+                rustTemplate("Ok(#{SdkBody}::from(out))", *codegenScope)
             }
         }
     }
 
-    override fun operationSerializer(operationShape: OperationShape): RuntimeType? {
-        val fnName = symbolProvider.serializeFunctionName(operationShape)
-        val inputShape = operationShape.inputShape(model)
-        return RuntimeType.forInlineFun(fnName, "operation_ser") { writer ->
+    private fun RustWriter.serializeStructure(context: Context<StructureShape>, topLevel: Boolean) {
+        val fnName = symbolProvider.serializeFunctionName(context.shape)
+        val structureSymbol = symbolProvider.toSymbol(context.shape)
+        val structureSerializer = RuntimeType.forInlineFun(fnName, "query_ser") { writer ->
+            val writerMut = when (topLevel) {
+                true -> ""
+                else -> "mut "
+            }
+            val writerType = when (topLevel) {
+                true -> "&mut #{QueryWriter}"
+                false -> "#{QueryValueWriter}"
+            }
+            Attribute.AllowUnusedMut.render(writer)
```

Comment:
I've done worse, but in the past I've dealt with this by having `serializeStructureInner`. When it's used for operations, we inline `serializeStructureInner`.

Serialize structure sets up a function signature then calls `serializeStructureInner`

Reply:
This is fixed.